pulsar-transformations/src/main/java/com/datastax/oss/pulsar/functions/transforms/FlattenStep.java

```diff
@@ -0,0 +1,165 @@
+/*
+ * Copyright DataStax, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.datastax.oss.pulsar.functions.transforms;
+
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Optional;
+import java.util.stream.Collectors;
+import lombok.RequiredArgsConstructor;
+import lombok.Value;
+import org.apache.avro.generic.GenericData;
+import org.apache.avro.generic.GenericRecord;
+import org.apache.pulsar.client.api.Schema;
+import org.apache.pulsar.common.schema.SchemaType;
+
+@RequiredArgsConstructor
+public class FlattenStep implements TransformStep {
+  // TODO: Cache schema to flattened schema
+  // TODO: Microbenchmark the flatten algorithm for performance optimization
+  // TODO: Validate flatten delimiter
+  // TODO: Add integration test
+
+  private static final String DEFAULT_DELIMITER = "_"; // '.' in not valid in AVRO field names
+
+  private final Optional<String> delimiter;
```

Comment:
Some people would say that `Optional` was only meant to be a type for return of methods and not for class attributes.

Reply:
makes sense, let me find a better way to handle it.